ntoskrnl/mm/marea.c

```diff
@@ -314,16 +314,19 @@ MmFreeMemoryArea(
             BOOLEAN Dirty = FALSE;
             SWAPENTRY SwapEntry = 0;
             PFN_NUMBER Page = 0;
+            BOOLEAN DoFree;
 
             if (MmIsPageSwapEntry(Process, (PVOID)Address))
             {
                 MmDeletePageFileMapping(Process, (PVOID)Address, &SwapEntry);
+                /* We'll have to do some cleanup when we're on the page file */
+                DoFree = TRUE;
             }
             else
             {
-                MmDeleteVirtualMapping(Process, (PVOID)Address, &Dirty, &Page);
+                DoFree = MmDeleteVirtualMapping(Process, (PVOID)Address, &Dirty, &Page);
             }
-            if (FreePage != NULL)
+            if (DoFree && (FreePage != NULL))
```

Comment:
Here, contrary to in the rmap.c / section.c cases, is it expected that `DoFree` can be FALSE, without this situation being critical?

Reply:
Yes. In this case, it's deleting a whole chunk of the address space, not knowing whether there's actually a page mapped or not. In the other cases, it's unmapping a specific address where we know there is a mapping.